### ISSUE TITLE
Texture size too large causes invisible text

### ISSUE THREAD

**Issue by cropleyb** [kivy/kivy#2119]
I have a large amount of scrollable text which was in a single Label. This works fine on my development machine (OS X) because I have the screen resolution set quite low (my eyesight isn't great). On iOS, this text disappeared. The reason: The text required a texture size that was too large. The workaround was to create a GridLayout containing several Labels. This works fine, but IMO it shouldn't be necessary.
Ideally, I'd like the texture to (transparently to the developer) create a composite texture.
If that is too difficult then it should at least raise an exception so that the developer can figure out what is going on.
Thanks,
Bruce
##

**Comment by abhigenie92** [comment]
Same issue 
http://stackoverflow.com/questions/32412903/blank-text-output-on-android-phone

**Comment by melassa** [comment]
There is a way to understand when the texture size limit is reached?

**Comment by younis-jd** [comment]
I have a Lable inside a ScrollView and I have the same issue. I'm running my Kivy app on a Raspberry Pi 3. Is there a way to know the limit. It seems to have a different limit everytime the app is running. Because I can see the view turns blank at different amount of text. Thank you :)

**Comment by matham** [comment]
I think generally if you reach the maximum size in your app, you should probably change your app to break up text in smaller labels.

The texture size I believe is fixed for a particular system.

**Comment by melassa** [comment]
I divide my text but the problem is to know the limit for a particular android phone, tablet, version

**Comment by dessant** [comment]
@melassa, the resolution of the display is guaranteed to work as texture size, if you need a safe limit without querying gl features.

**Comment by melassa** [comment]
Ok.

**Comment by uxmatthew** [comment]
I'm also running into this, would prefer not to have to code more to break it up, but will do for now I guess.

**Comment by chintal** [comment]
This is something of a nasty problem, in my opinion. It's especially bad because it isn't even close to obvious that this can be an issue. I've spent the last few hours playing with size, texture_size, text_size  and hints with very strange combinations of results. This is the first time I'm noticing that there is a maximum texture size and consequently a maximum text size on labels, and the first piece of code was actually fine otherwise. Label docs (and any other widgets that could conceivably be hit by this) should mention this.

**Comment by matham** [comment]
Unfortunately, this is one of the things that people new to kivy must wade through and hit once before they learn about it. I'll create a label for these kinds of issues so we can document it in a faq e.g. But yeah, the docs should mention it.

**Comment by KeyWeeUsr** [comment]
@abhigenie92 the link is dead
@melassa, @chintal 

Notice in the console this (my max size):

> [INFO   ] [GL          ] Texture max size <16384>

```python
from kivy.base import runTouchApp
from kivy.lang import Builder
runTouchApp(Builder.load_string('''
#:import T kivy.graphics.opengl.GL_MAX_TEXTURE_SIZE
#:import G kivy.graphics.opengl.glGetIntegerv
TextInput:
    multiline: False
    on_text:
        print(any([s >= G(T)[0] for lb in self._lines_labels for s in lb.size]))
'''))
```

Pretty much you shouldn't be worried about the texture height (unless you use a silly font size), therefore it might be shortened to this:

        print(any([lb.size[0] >= G(T)[0] for lb in self._lines_labels]))

which is a simple check whether your texture is displayed or not. I think the caveat is when e.g. the (separate or integrated) GPU's memory is full, then this check won't matter and the texture might not be displayed at all or it could just crash or smth (haven't encountered such case in Kivy, can't say what happens).

With this check or perhaps with an adjusted one with an index if the `Texture` you can detect which one is just full and either move the text to the next line/widget/texture yourself or only limit the widget in this way:

```python
from kivy.base import runTouchApp
from kivy.lang import Builder
from kivy.uix.textinput import TextInput
from kivy.graphics.opengl import GL_MAX_TEXTURE_SIZE
from kivy.graphics.opengl import glGetIntegerv


class My(TextInput):
    def insert_text(self, *args, **kwargs):
        textures = self._lines_labels
        assert len(textures) == 1
        # assuming singleline mode, no silly \n inserting, etc

        max_size = glGetIntegerv(GL_MAX_TEXTURE_SIZE)[0]
        if textures[0].width >= max_size:
            print('GL_MAX_TEXTURE_SIZE({}) exceeded!'.format(max_size))
            # don't insert any text
            return
        super(My, self).insert_text(*args, **kwargs)


runTouchApp(My(multiline=False))
```

Try inserting character by character or Copy&Paste long string. Obviously if the `Texture`'s size is less than the max and you insert more than the difference, the text will be just a black box, but the behavior is upon you i.e. whether you insert by chunks or by characters.

**Comment by charlesbaynham** [comment]
Just to mention that yep, I ran into this too. After an hour or so of playing with texture_size, etc I finally came across this thread. It'd be great if the user didn't have to worry about this, but in the meantime the docs (probably for Label) really should mention it somewhere.

**Comment by PremKolar** [comment]
same problem :(

**Comment by NikitaAkimov** [comment]
Unfortunately, this problem still exists at the moment.

**Comment by matham** [comment]
I think perhaps the best solution is to raise an exception when this happens, rather than silently not work to make it obvious to the user what they need to fix.

The actual issue is not likely to be fixed because this a gl limitation, and in a performant app, it probably shouldn't create these giant textures at once anyway. So this is not something kivy should create workarounds for but leave it up to app dev to split their textures up. But we should make it clear somehow to the user when this happens.

A PR raising an error would be welcome to create a discussion around this.

**Comment by Codingboy** [comment]
I am writing an epub reader and wanted to display a chapter in a label which is in a scrollview. Looks like I have to code what I would expect a UI framework to do for me.

**Comment by CoderCoderCoder** [comment]
Same issue :-(. Please provide at least some workaround code, how to deal with it properly. E.g. splitting text on multiple label automatically, if the texture size is bigger than the max GPU limit.

**Comment by CoderCoderCoder** [comment]
OK, to all suffering the same issue, here is the workaround.
First of all, there is also the error in the output "UNSUPPORTED (log once): POSSIBLE ISSUE: unit 0 GLD_TEXTURE_INDEX_2D is unloadable and bound to sampler type (Float) - using zero texture because texture unloadable", which indicates that the texture is too large to be loaded.

Now the actual workaround:

from kivy.graphics.opengl import glGetIntegerv, GL_MAX_TEXTURE_SIZE
MAX_TEXTURE_SIZE = glGetIntegerv(GL_MAX_TEXTURE_SIZE)[0]

def adapt_text_size(self, instance, value):
    if value[0] >= MAX_TEXTURE_SIZE:
        instance.text_size[0] = MAX_TEXTURE_SIZE - 1
    if value[1] >= MAX_TEXTURE_SIZE:
        instance.text_size[1] = MAX_TEXTURE_SIZE - 1
    instance.canvas.ask_update() 

and you need to bind it on any UI element (in my case MDLabel), which you suspect can grow too big. 

label.bind(texture_size=self.adapt_text_size) 

That's it!

**Comment by chintal** [comment]
If you have a use case which actually needs a large texture size, you can also stitch multiple smaller labels together into a layout and use that instead. An example of this is in https://github.com/ebs-universe/kivy_garden.ebs.marquee/blob/main/kivy_garden/ebs/marquee/simple.py , specifically the text setter. While this won't be usable as is for the general case, the same approach can be adapted for your use case. 

Be careful when doing this, though. The implementation above is also pretty naive, and if you try to push too much text in, you will end up with proportionally increased memory use.

**Comment by makeasnek** [comment]
Please do not use bountysource. Many devs have had trouble getting paid there. You can check out this lemmy community as an alternative https://lemmy.ml/c/bugbounties

For statements from devs who have been unable to cash out from bountysource see:
https://github.com/bountysource/core/issues

**Comment by Julian-O** [comment]
@makeasnek: Yes, BountySource is not recommended any more. It has been removed from our docs #8447.

However, it remains locked into the history of old issues, unless someone writes a bot to clear it out.
